saleor/checkout/utils.py

```diff
@@ -149,13 +149,22 @@ def add_variants_to_checkout(checkout, variants, quantities, channel_slug):
         if not product_channel_listing or not product_channel_listing.is_published:
             raise ProductNotPublished()
 
-    # create checkout lines
-    lines = []
+    lines = {line.variant.pk: line for line in checkout.lines.all()}
+    to_create = []
     for variant, quantity in zip(variants, quantities):
-        lines.append(
-            CheckoutLine(checkout=checkout, variant=variant, quantity=quantity)
-        )
-    checkout.lines.bulk_create(lines)
+        if variant.pk in lines:
+            line = lines[variant.pk]
+            if quantity > 0:
+                line.quantity = quantity
```

Comment:
Doesn't make sense to call bulk here also? Same question for `delete`

Reply:
Yep, made these bulk as well.